### PR TITLE
LibMarkdown: Add support for different types of blocks in lists

### PR DESCRIPTION
**Pull request by Petelliott**
- lists can contain different blocks (lists, code blocks, etc) instead of just text
- [loose/tight lists](https://spec.commonmark.org/0.30/#loose)
- spaces before list items
- non-1 start numbers for ordered lists

`TestCommonmark` pass rate:
227/652 -> 253/652

### CODE REVIEW

File: Userland/Libraries/LibMarkdown/List.h [hunk 1]
```diff
@@ -15,21 +15,23 @@ namespace Markdown {
...
-    virtual String render_to_html() const override;
+    virtual String render_to_html(bool tight = false) const override;
     virtual String render_for_terminal(size_t view_width = 0) const override;
 
     static OwnPtr<List> parse(LineIterator& lines);
 
 private:
     Vector<OwnPtr<ContainerBlock>> m_items;
     bool m_is_ordered { false };
+    bool m_is_tight;
```
Comment (alimpfard) [inline Userland/Libraries/LibMarkdown/List.h:35]: Please default-initialise values of primitive types.

File: Userland/Libraries/LibMarkdown/List.h [hunk 2]
```diff
@@ -32,6 +33,7 @@ class List final : public Block {
     Vector<OwnPtr<ContainerBlock>> m_items;
     bool m_is_ordered { false };
     bool m_is_tight;
+    size_t m_start_number;
```
Comment (alimpfard) [inline Userland/Libraries/LibMarkdown/List.h:36]: Please default-initialise values of primitive types.

File: Userland/Libraries/LibMarkdown/HorizontalRule.h [hunk 3]
```diff
@@ -21,7 +21,7 @@ class HorizontalRule final : public Block {
     }
     virtual ~HorizontalRule() override { }
 
-    virtual String render_to_html() const override;
+    virtual String render_to_html(bool tight = false) const override;
```
Comment (alimpfard) [inline Userland/Libraries/LibMarkdown/HorizontalRule.h:24]: Some of the overrides for this function have default values for `tight`, and some don't.
Please pick one and go with it?

File: Userland/Libraries/LibMarkdown/List.cpp [hunk 4]
```diff
@@ -38,96 +47,95 @@ String List::render_for_terminal(size_t) const
...
 
         bool appears_ordered = false;
-        for (size_t i = 0; i < 10 && i < line.length(); i++) {
+        for (size_t i = offset; i < 10 && i < line.length(); i++) {
             char ch = line[i];
             if ('0' <= ch && ch <= '9')
                 continue;
             if (ch == '.' || ch == ')')
                 if (i + 1 < line.length() && line[i + 1] == ' ') {
+                    auto maybe_start_number = line.substring_view(offset, i - offset).to_uint<size_t>();
```
Comment (alimpfard) [inline Userland/Libraries/LibMarkdown/List.cpp:94]: Also, looks like this specialisation of `to_uint` doesn't exist.
These two silly declarations were probably meant to be marked `unsigned` (in StringView.cpp:230):

```c++
template Optional<long> StringView::to_uint() const;
template Optional<long long> StringView::to_uint() const;
```

File: Userland/Libraries/LibMarkdown/HorizontalRule.h (hunk 3 above)
Comment (Petelliott) [reply Userland/Libraries/LibMarkdown/HorizontalRule.h:24]: I went with all default values

File: Userland/Libraries/LibMarkdown/List.cpp (hunk 4 above)
Comment (Petelliott) [reply Userland/Libraries/LibMarkdown/List.cpp:94]: This is very weird, see my commit message for the last commit.

**Review by alimpfard — COMMENTED** [review]
sizeof(long long) == sizeof(long) doesn't make much sense on i686, are you sure that's the case?
I'd expect long to be 32 bits, and long long to be 64 bits long

**Comment by Petelliott** [comment]
> sizeof(long long) == sizeof(long) doesn't make much sense on i686, are you sure that's the case? I'd expect long to be 32 bits, and long long to be 64 bits long

You're right. that's inaccurate. I was accidentally looking at output from the lagom build.

The real issue is that on i686 `u64 == long long` but on i686 `u64 == long` so which instantiation is missing on each arch is different, but we need both. all the other `u...` types are the same, so they can remain. my change is still correct, but I will update the description.

File: Userland/Libraries/LibMarkdown/List.cpp (hunk 4 above)
Comment (Petelliott) [reply Userland/Libraries/LibMarkdown/List.cpp:94]: also the integer instantiations are actually used in libline, so i'm leaving them.

File: Userland/Libraries/LibMarkdown/ContainerBlock.cpp [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    for (auto& block : m_blocks) {
+        auto s = block.render_for_terminal(view_width);
+        builder.append(s);
+    }
+
+    return builder.build();
+}
+
+template<typename BlockType>
+static bool helper(Vector<StringView>::ConstIterator& lines, NonnullOwnPtrVector<Block>& blocks)
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibMarkdown/ContainerBlock.cpp]: Could this be named better? For example `parse_to_blocks`?

Comment (Petelliott) [reply Userland/Libraries/LibMarkdown/ContainerBlock.cpp]: `try_parse_block` would probably be a more accurate name. I copied this from document.h but I'll update it.

**Comment by alimpfard** [comment]
Let's get this in.

**Pull request merged**